test/functional/feature_assumeutxo.py

```diff
@@ -229,6 +229,15 @@ def test_snapshot_in_a_divergent_chain(self, dump_output_path):
         self.log.info("Ensuring background validation finishes")
         self.wait_until(lambda: len(n3.getchainstates()['chainstates']) == 1)
 
+        self.log.info("Check importing a snapshot where current chain-tip is not an ancestor or a descendant of the snapshot block and has more work")
+        # Generate a divergent chain in n4 that has more work than the snapshot
+        # This covers the scenario where the snapshot block is not on the most-work chain
+        self.generate(n4, nblocks=101, sync_fun=self.no_op)
+        assert_equal(n4.getblockcount(), SNAPSHOT_BASE_HEIGHT + 1)
+        # Import the snapshot and assert its failure
+        with n4.assert_debug_log(expected_msgs=["[snapshot] activation failed - work does not exceed active chainstate"]):
```

Comment:
This goes back to some of my earlier comments but I think it got lost because there were multiple things discussed at the same time: I think this is not the error we want to hit in this case (as I interpret the TODO) and I think we will not hit it in the future either, because once #30320 is merged this will hit the error message added there and not this one. This error message here mentions "work" but this is the actual work of the active chainstate, not some headers that the node has received. I.e. this is the case where the node has basically already synced when loading the utxo set. And that makes sense because here the node loading the snapshot has generated the chain itself, so it has already synced too. So I am not sure that this case here adds anything new compared to the existing case for this scenario in `test_snapshot_with_less_work()`. And I tend to think that actually #30320 completes the TODOs as I understand them and this commit here is not needed in addition to #30320 and `test_snapshot_with_less_work()`.

cc @mzumsande - What do you think?

Reply:
Thanks @fjahr . I'm open to deleting the third commit. I mentioned this [earlier](https://github.com/bitcoin/bitcoin/pull/29996#issuecomment-2086782948) but kept it because I thought my scenario might be somehow different due to the divergent chain.

Your explanation helped me understand the concept of the headers chain, and I think you're right that #30320 may cover the remaining TODOs. I interpreted the part about `"...loading a snapshot when the current chain tip is: [...]"` as referring to the chainstate of the node importing the snapshot, not the network's chain tip.

I'll wait for @mzumsande's input to confirm before deleting the third commit if needed. I'll also address your new feedback.

Reply:
I think that "chain tip" usually refers to blocks we have connected to a chainstate (so headers-only would not qualify), so I would suspect that the original intention of the TODO was just what the third commit implements.

But I agree with @fjahr that it doesn't really add too much to `test_snapshot_with_less_work` - less work is less work, no matter if the chain is divergent or not. Plus, I agree that if #30320 gets merged it would become a special case of that, because a chain tip with more work than the snapshot requires a header with more work (but not vice versa).

These are just my 2 cents, will review the PR in more detail soon.

Unrelated: @alfonsoromanz Thanks for the mention, but could you remove the "@" from my name in the PR description - if it'd get merged with it, I'll get notifications whenever some altcoin fork cherrypicks that commit later.

Reply:
Thanks @mzumsande. Makes sense.

I removed the third commit since there is agreement that it is redundant with `test_snapshot_with_less_work()`.

> Unrelated: @alfonsoromanz Thanks for the mention, but could you remove the "@" from my name in the PR description - if it'd get merged with it, I'll get notifications whenever some altcoin fork cherrypicks that commit later.

Sorry about that. The "@" was already removed by fanquake.